tonilopezmr/day 2/ChecksumPartOne.kt

```diff
@@ -0,0 +1,71 @@
+package day2
+
+import org.junit.Assert.*
+import org.junit.Test
+
+class ChecksumPartOne {
+
+  fun checksum(of: Array<Array<Int>>): Int {
+    return of.fold(emptyList<Int>()) { acc, a ->
+      acc.plus(maxMinusMin(a))
+    }.reduce { acc, i -> acc + i }
+  }
```

Comment:
This is nice but you could make this oneliner and save one iteration (or use `sum` instead of reduce). In my opinion, your second iteration is trivial so I would have done:

```kotlin
fun checksum(of: Array<Array<Int>>): Int = of.fold(0) { acc, row ->
  acc + maxMinusMin(row)
}
```

Reply:
This way instead of `O(2n)` we have `O(n)` if I am not mistaken.

Reply:
xD that's true

Reply:
Now that I think, you could even simplify more and change `fold(0) { ... }` by `reduce { ... }` because reduce starts with the head of the list as accumulator which is basically the same as `0 + head` which fold is making.

And If you just want a single parameter in the lambda function you could just do `of.sumBy(maxMinusMin)`